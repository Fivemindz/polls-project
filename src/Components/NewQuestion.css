.new-question{
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 100%;
  grid-area: mn;
}
.new-question h1 {
  width: 100%;
}
.new-question-form{
  border: 1px solid black;
  width: 80%;
  height: 50%;
  display: flex;
  flex-direction: column;
  align-items: center;
}
.question-area{
  display: flex;
  flex-direction: column;
  margin: 20px;
}
.question-area textarea{
  margin: 20px;
  width: 350px;
  font-size: large;
}
