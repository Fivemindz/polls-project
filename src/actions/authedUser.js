export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const RECEIVE_USER = "RECEIVE_USER";

export function setAuthedUser(user) {
  return {
    type: SET_AUTHED_USER,
    user,
  };
}

export function receiveUser(user) {
  return {
    type: RECEIVE_USER,
    user,
  };
}
