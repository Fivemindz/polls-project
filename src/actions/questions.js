import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { showLoading, hideLoading } from "react-redux-loading-bar";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const SAVE_QUESTION = "SAVE_QUESTIONS";
export const SAVE_QUESTION_ANSWER = "SAVE_QUESTION_ANSWER";

function addQuestion(question) {
  return {
    type: SAVE_QUESTION,
    question,
  };
}

function addQuestionAnswer(answer) {
  return {
    type: SAVE_QUESTION_ANSWER,
    answer,
  };
}

export function handleAddQuestion(optionOneText, optionTwoText) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    dispatch(showLoading());

    return saveQuestion({
      optionOneText,
      optionTwoText,
      author: authedUser,
    })
      .then((question) => {
        dispatch(addQuestion(question));
      })
      .then(() => dispatch(hideLoading()));
  };
}

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function handleAddQuestionAnswer(qid, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    dispatch(showLoading());

    return saveQuestionAnswer({
      authedUser: authedUser,
      qid,
      answer,
    })
      .then((answer) => dispatch(addQuestionAnswer(answer)))
      .then(() => dispatch(hideLoading()));
  };
}
