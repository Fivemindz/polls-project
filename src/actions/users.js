export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_QUESTION_ANSWER = "ADD_QUESTION_ANSWER";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}
