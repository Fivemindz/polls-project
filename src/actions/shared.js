import { getInitialData, loadUsers } from "../utils/api";
import { receiveUsers } from "./users";
import { addQuestion, addQuestionAnswer, receiveQuestions } from "./questions";
import { setAuthedUser } from "./authedUser";
import { showLoading, hideLoading } from "react-redux-loading-bar";
import { saveQuestion, saveQuestionAnswer } from "../utils/api";

export function handleInitialData(user) {
  return (dispatch) => {
    dispatch(showLoading());
    return getInitialData().then(({ users, questions }) => {
      dispatch(receiveUsers(users));
      dispatch(receiveQuestions(questions));
      dispatch(setAuthedUser(user));
      dispatch(hideLoading());
    });
  };
}

export function getUsers() {
  return async (dispatch) => {
    dispatch(showLoading());
    return loadUsers().then(({ users }) => {
      dispatch(receiveUsers(users));
      dispatch(hideLoading());
    });
  };
}

export function handleAddQuestion(optionOneText, optionTwoText) {
  return (dispatch, getState) => {
    const { authedUser } = getState();

    dispatch(showLoading());

    return saveQuestion({
      optionOneText,
      optionTwoText,
      author: authedUser,
    })
      .then((question) => {
        dispatch(addQuestion(question));
      })
      .then(() => dispatch(hideLoading()));
  };
}

export function handleAddQuestionAnswer(info) {
  return async (dispatch) => {
    dispatch(showLoading());
    return saveQuestionAnswer(info).then(async () => {
      await dispatch(addQuestionAnswer(info));
      dispatch(hideLoading());
    });
  };
}
