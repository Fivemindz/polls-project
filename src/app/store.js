import { configureStore } from "@reduxjs/toolkit";
import { authedUser as authedUserReducer } from "../reducers/authedUser";
import { questions as questionsReducer } from "../reducers/questions";
import { users as usersReducer } from "../reducers/users";
import logger from "../middleware/logger";
import thunk from "redux-thunk";

export const store = configureStore({
  reducer: {
    authedUser: authedUserReducer,
    questions: questionsReducer,
    users: usersReducer,
  },
  middleware: [thunk],
});
