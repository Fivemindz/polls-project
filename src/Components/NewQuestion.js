import React, { useRef } from "react";
import { connect } from "react-redux";
import "./NewQuestion.css";
import { handleAddQuestion } from "../actions/shared";
import { useNavigate } from "react-router-dom";

const NewQuestion = (props) => {
  const { authedUser, userPic, dispatch } = props;

  const navigate = useNavigate();

  const optionOneText = useRef("");
  const optionTwoText = useRef("");

  const handleSubmit = async (e) => {
    e.preventDefault();
    try {
      await dispatch(
        handleAddQuestion(optionOneText.current, optionTwoText.current)
      );
    } catch {
      return;
    }
    optionOneText.current = "";
    optionTwoText.current = "";

    navigate("/");
  };

  return (
    <div className="new-question">
      <h2>New Poll by {authedUser} </h2>
      <img src={userPic} alt="avatar" className="avatar" />
      <span></span>
      <form
        className="new-question-form"
        onSubmit={handleSubmit}
        data-testid="new-question-form"
      >
        <h3>Would You Rather?</h3>
        <div className="question-area">
          <label>Answer 1</label>
          <textarea
            id="optionOneText"
            onChange={(e) => (optionOneText.current = e.target.value)}
            data-testid="question-text-1"
          />
          <label>Answer 2</label>
          <textarea
            id="optionTwoText"
            onChange={(e) => (optionTwoText.current = e.target.value)}
            data-testid="question-text-2"
          />
        </div>
        <button
          className="question-submit-btn"
          type="submit"
          disabled={
            optionOneText.current.length !== 0 ||
            optionTwoText.current.length !== 0
          }
          data-testid="question-submit-btn"
        >
          Submit
        </button>
      </form>
    </div>
  );
};

const mapStateToProps = ({ authedUser, users }) => ({
  authedUser,
  userPic: users[authedUser].avatarURL,
});

export default connect(mapStateToProps)(NewQuestion);
