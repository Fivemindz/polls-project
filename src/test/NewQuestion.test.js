import { fireEvent, render, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "./store";
import { BrowserRouter as Router } from "react-router-dom";
import NewQuestion from "../Components/NewQuestion";
import { handleInitialData } from "../actions/shared";

describe("Test NewQuestion component ", () => {
  beforeAll(async () => {
    await store.dispatch(handleInitialData("tylermcginnis"));
  });

  it("Snapshot: Renders the NewQuesion form", async () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <NewQuestion />
        </Router>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it("Test question has been added after all expected fields are populated", async () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <NewQuestion />
        </Router>
      </Provider>
    );

    const text1 = component.getByTestId("question-text-1");
    const text2 = component.getByTestId("question-text-2");
    const btn = component.getByTestId("question-submit-btn");
    fireEvent.change(text1, { target: { value: "option 1" } });
    fireEvent.change(text2, { target: { value: "option 2" } });
    fireEvent.click(btn);

    waitFor(async () => {
      await component.getByTestId("dash");
      const dash = component.getByTestId("dash");
      expect(dash).toBeInTheDocument();
    });
  });

  it("Test question will not add if fields are not populated", async () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <NewQuestion />
        </Router>
      </Provider>
    );

    const btn = component.getByTestId("question-submit-btn");
    fireEvent.click(btn);
    waitFor(async () => {
      await component.getByTestId("dash");
      const questionForm = component.getByTestId("new-question-form");
      expect(questionForm).toBeInTheDocument();
    });
  });
});
