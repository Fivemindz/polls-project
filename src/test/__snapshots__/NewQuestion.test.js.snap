// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test NewQuestion component  Snapshot: Renders the NewQuesion form 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="new-question"
      >
        <h2>
          New Poll by 
          tylermcginnis
           
        </h2>
        <img
          alt="avatar"
          class="avatar"
          src="./avatars/tyler.jpg"
        />
        <span />
        <form
          class="new-question-form"
          data-testid="new-question-form"
        >
          <h3>
            Would You Rather?
          </h3>
          <div
            class="question-area"
          >
            <label>
              Answer 1
            </label>
            <textarea
              data-testid="question-text-1"
              id="optionOneText"
            />
            <label>
              Answer 2
            </label>
            <textarea
              data-testid="question-text-2"
              id="optionTwoText"
            />
          </div>
          <button
            class="question-submit-btn"
            data-testid="question-submit-btn"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="new-question"
    >
      <h2>
        New Poll by 
        tylermcginnis
         
      </h2>
      <img
        alt="avatar"
        class="avatar"
        src="./avatars/tyler.jpg"
      />
      <span />
      <form
        class="new-question-form"
        data-testid="new-question-form"
      >
        <h3>
          Would You Rather?
        </h3>
        <div
          class="question-area"
        >
          <label>
            Answer 1
          </label>
          <textarea
            data-testid="question-text-1"
            id="optionOneText"
          />
          <label>
            Answer 2
          </label>
          <textarea
            data-testid="question-text-2"
            id="optionTwoText"
          />
        </div>
        <button
          class="question-submit-btn"
          data-testid="question-submit-btn"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
